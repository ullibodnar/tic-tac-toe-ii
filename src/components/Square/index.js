import React from 'react'
import styled from 'styled-components'
import { isUndefined } from 'ramda-adjunct'

const StyledSquare = styled.div`
  border-color: #191919;
  border-style: solid;
  border-width: 0 ${({ index }) => (index % 3 === 2 ? 0 : '3px')}
    ${({ index }) => (index < 6 ? '3px' : 0)} 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  {/*text-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);*/}
`
StyledSquare.defaultName = 'StyledSquare'

const SquarePlayed = StyledSquare.extend`
  color: ${({ player }) => (player === 'x' ? '#FF8080' : '#80FFFF')};

`

const BlockSuggestion = StyledSquare.extend`
  border: 1px dashed red;
`

const SquareLost = StyledSquare.extend`
  color: hsla(0, 0%, 90%, 1);
`

const SquarePlayable = StyledSquare.extend`
  cursor: pointer;
`

export default function Square ({
  handleClick,
  index,
  isWinningSquare,
  player
}) {
  if (isUndefined(isWinningSquare)) {
    return isUndefined(player)
      ? <SquarePlayable index={index} onClick={handleClick} />
      : <SquarePlayed index={index} player={player}>{player}</SquarePlayed>
  }

  if (isUndefined(player)) {
    return <StyledSquare index={index} />
  }

  return isWinningSquare
    ? <SquarePlayed index={index} player={player}>{player}</SquarePlayed>
    : <SquareLost index={index} player={player}>{player}</SquareLost>
}
