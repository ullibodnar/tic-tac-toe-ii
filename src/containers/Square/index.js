import { connect } from 'react-redux'
import { contains } from 'ramda'
import { isNotEmpty } from 'ramda-adjunct'

import Square from '../../components/Square'
import { getMoves, getWinningSquares, squareClicked } from '../../state'
import { getPlayer } from '../../utilities'

function mapStateToProps (state, { index }) {
  const moves = getMoves(state)
  const winners = getWinningSquares(state) || []
  const gameIsOver = isNotEmpty(winners)
  const player = getPlayer(index, moves)

  return gameIsOver
    ? { player, isWinningSquare: contains(index, winners) }
    : { player }
}

function mapDispatchToProps (dispatch, { index }) {
  return {
    handleClick: () => dispatch(squareClicked(index))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Square)
