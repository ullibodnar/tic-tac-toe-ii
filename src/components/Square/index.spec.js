import React from 'react'
import { shallow } from 'enzyme'

import Square from '.'

describe('components:Square', () => {
  it('renders the Square with the proper cursor if a click handler is provided', () => {
    expect(
      toJson(shallow(<Square handleClick={() => null} index={0} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player O in the top left square', () => {
    expect(
      toJson(shallow(<Square player='o' index={0} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player X in the top left square', () => {
    expect(
      toJson(shallow(<Square player='x' index={0} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player X in the top right square', () => {
    expect(
      toJson(shallow(<Square player='x' index={2} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player X in the bottom left square', () => {
    expect(
      toJson(shallow(<Square player='x' index={6} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player X in the bottom right square', () => {
    expect(
      toJson(shallow(<Square player='x' index={8} />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player X win', () => {
    expect(
      toJson(shallow(<Square player='x' index={8} isWinningSquare />).dive())
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for player O loss', () => {
    expect(
      toJson(
        shallow(<Square player='o' index={8} isWinningSquare={false} />).dive()
      )
    ).toMatchSnapshot()
  })

  it('renders the Square with the proper styles for an unplayed square after game over', () => {
    expect(
      toJson(shallow(<Square index={4} isWinningSquare={false} />).dive())
    ).toMatchSnapshot()
  })
})
