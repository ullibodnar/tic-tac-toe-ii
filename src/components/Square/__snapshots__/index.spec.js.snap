// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:Square renders the Square with the proper cursor if a click handler is provided 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 3px 3px 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  cursor: pointer;
}

<div
  className="c0"
  onClick={[Function]}
/>
`;

exports[`components:Square renders the Square with the proper styles for an unplayed square after game over 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 3px 3px 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
}

<div
  className="c0"
/>
`;

exports[`components:Square renders the Square with the proper styles for player O in the top left square 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 3px 3px 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #e2e1e0;
}

<div
  className="c0"
>
  o
</div>
`;

exports[`components:Square renders the Square with the proper styles for player O loss 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 0 0 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: hsla(0,0%,90%,1);
}

<div
  className="c0"
>
  o
</div>
`;

exports[`components:Square renders the Square with the proper styles for player X in the bottom left square 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 3px 0 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #fc3645;
}

<div
  className="c0"
>
  x
</div>
`;

exports[`components:Square renders the Square with the proper styles for player X in the bottom right square 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 0 0 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #fc3645;
}

<div
  className="c0"
>
  x
</div>
`;

exports[`components:Square renders the Square with the proper styles for player X in the top left square 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 3px 3px 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #fc3645;
}

<div
  className="c0"
>
  x
</div>
`;

exports[`components:Square renders the Square with the proper styles for player X in the top right square 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 0 3px 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #fc3645;
}

<div
  className="c0"
>
  x
</div>
`;

exports[`components:Square renders the Square with the proper styles for player X win 1`] = `
.c0 {
  border-color: #191919;
  border-style: solid;
  border-width: 0 0 0 0;
  cursor: default;
  font-size: 16vh;
  font-weight: bold;
  line-height: 20vh;
  text-align: center;
  text-transform: uppercase;
  text-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  color: #fc3645;
}

<div
  className="c0"
>
  x
</div>
`;
